### PR TITLE
Uplift next trait solver to `rustc_next_trait_solver`

### PR DESCRIPTION
**Pull request by compiler-errors**
🎉 

There's so many FIXMEs! Sorry! Ideally this merges with the FIXMEs and we track and squash them over the near future.

Also, this still doesn't build on anything other than rustc. I still need to fix `feature = "nightly"` in `rustc_type_ir`, and remove and fix all the nightly feature usage in the new trait solver (notably: let-chains).

Also, sorry @lcnr I know you asked for me to separate the commit where we `mv rustc_trait_selection/solve/... rustc_next_trait_solver/solve/...`, but I had already done all the work by that point. Luckily, `git` understands the file moves so it should still be relatively reviewable.

If this is still very difficult to review, then I can do some rebasing magic to try to separate this out. Please let me know!

r? lcnr

### CODE REVIEW

File: compiler/rustc_infer/src/infer/mod.rs [hunk 1]
```diff
@@ -334,149 +334,6 @@ pub struct InferCtxt<'tcx> {
     pub obligation_inspector: Cell<Option<ObligationInspector<'tcx>>>,
 }
 
-impl<'tcx> ty::InferCtxtLike for InferCtxt<'tcx> {
```
Comment (compiler-errors) [inline compiler/rustc_infer/src/infer/mod.rs:337]: This was moved to `rustc_trait_selection/src/solve/infcx.rs`

File: compiler/rustc_infer/src/infer/mod.rs [hunk 2]
```diff
@@ -830,6 +687,10 @@ impl<'tcx> InferCtxt<'tcx> {
         self.tcx.dcx()
     }
 
+    pub fn defining_opaque_types(&self) -> &'tcx ty::List<LocalDefId> {
```
Comment (compiler-errors) [inline compiler/rustc_infer/src/infer/mod.rs:690]: Since `defining_opaque_types` is a private field on `InferCtxt`, and I've chosen to expose only one trait (`SolverDelegate`) and not two (`InferCtxtLike`), we gotta make it possible to reference this in the impl in `rustc_trait_solver`.

File: compiler/rustc_middle/src/traits/solve/cache.rs [hunk 3]
```diff
@@ -48,7 +42,7 @@ impl<'tcx> EvaluationCache<'tcx> {
         if cfg!(debug_assertions) {
             drop(map);
             let expected = CacheData { result, proof_tree, additional_depth, encountered_overflow };
-            let actual = self.get(tcx, key, [], Limit(additional_depth));
```
Comment (compiler-errors) [inline compiler/rustc_middle/src/traits/solve/cache.rs:51]: `Limit` is defined in `rustc_session`, and I was *kinda* too lazy to find a good ancestor crate to put `Limit` in...

Could fix it as a follow-up or fix it in this PR if necessary.

File: compiler/rustc_middle/src/ty/context.rs [hunk 4]
```diff
@@ -89,62 +90,87 @@ use std::ops::{Bound, Deref};
 impl<'tcx> Interner for TyCtxt<'tcx> {
     type DefId = DefId;
     type LocalDefId = LocalDefId;
-    type AdtDef = ty::AdtDef<'tcx>;
-
     type GenericArgs = ty::GenericArgsRef<'tcx>;
+
```
Comment (compiler-errors) [inline compiler/rustc_middle/src/ty/context.rs:94]: Sorry, whenever I use rust-analyzer to sort items by definition, the spacing gets all fucked up.

File: compiler/rustc_next_trait_solver/src/solve/assembly/mod.rs [hunk 5]
```diff
@@ -301,16 +302,14 @@ impl<'tcx> EvalCtxt<'_, InferCtxt<'tcx>> {
 
         let mut candidates = vec![];
 
-        self.assemble_non_blanket_impl_candidates(goal, &mut candidates);
+        self.assemble_impl_candidates(goal, &mut candidates);
```
Comment (compiler-errors) [inline compiler/rustc_next_trait_solver/src/solve/assembly/mod.rs:305]: This consolidates `assemble_{blanket,non_blanket}_impl_candidates`. See the definition of `for_each_relevant_impl` that I added to `Interner`.

File: compiler/rustc_next_trait_solver/src/solve/assembly/structural_traits.rs [hunk 6]
```diff
@@ -3,26 +3,24 @@
...
+use tracing::instrument;
 
-use crate::solve::EvalCtxt;
+use crate::infcx::SolverDelegate;
+use crate::solve::{EvalCtxt, Goal, NoSolution};
 
 // Calculates the constituent types of a type for `auto trait` purposes.
-//
-// For types with an "existential" binder, i.e. coroutine witnesses, we also
-// instantiate the binder with placeholders eagerly.
```
Comment (compiler-errors) [inline compiler/rustc_next_trait_solver/src/solve/assembly/structural_traits.rs:18]: This is not true anymore lol

File: compiler/rustc_next_trait_solver/src/solve/eval_ctxt/canonical.rs [hunk 7]
```diff
@@ -102,8 +97,8 @@ impl<'tcx> EvalCtxt<'_, InferCtxt<'tcx>> {
 
         // We only check for leaks from universes which were entered inside
         // of the query.
-        self.infcx.leak_check(self.max_input_universe, None).map_err(|e| {
-            trace!(?e, "failed the leak check");
+        self.infcx.leak_check(self.max_input_universe).map_err(|NoSolution| {
```
Comment (compiler-errors) [inline compiler/rustc_next_trait_solver/src/solve/eval_ctxt/canonical.rs:100]: Actually, `leak_check` returns `TypeError` which I *did* previously uplift. I changed this without thinking -- could leave a FIXME to restore it.

File: compiler/rustc_next_trait_solver/src/solve/eval_ctxt/canonical.rs [hunk 8]
```diff
@@ -189,26 +184,7 @@ impl<'tcx> EvalCtxt<'_, InferCtxt<'tcx>> {
...
-                    )
-                });
-            assert_eq!(member_constraints, vec![]);
-            let mut seen = FxHashSet::default();
-            outlives
-                .into_iter()
-                .filter(|(outlives, _)| seen.insert(*outlives))
-                .map(|(outlives, _origin)| outlives)
-                .collect()
+            self.infcx.make_deduplicated_outlives_constraints()
```
Comment (compiler-errors) [inline compiler/rustc_next_trait_solver/src/solve/eval_ctxt/canonical.rs:187]: Pushed all of the logic of iterating the constraints and deduplicating them into the infcx.

Comment (compiler-errors) [reply compiler/rustc_next_trait_solver/src/solve/eval_ctxt/canonical.rs:187]: Could make it just return a non-deduplicated list and then deduplicate it here though...?

File: compiler/rustc_next_trait_solver/src/solve/eval_ctxt/canonical.rs [hunk 9]
```diff
@@ -425,45 +396,33 @@ where
...
-    for i in orig_values.len()..state.value.var_values.len() {
-        let unconstrained = match state.value.var_values.var_values[i].unpack() {
-            ty::GenericArgKind::Lifetime(_) => {
-                infcx.next_region_var(RegionVariableOrigin::MiscVariable(span)).into()
-            }
-            ty::GenericArgKind::Type(_) => infcx.next_ty_var(span).into(),
-            ty::GenericArgKind::Const(_) => infcx.next_const_var(span).into(),
-        };
-
+    for &arg in &state.value.var_values.var_values[orig_values.len()..state.value.var_values.len()]
```
Comment (compiler-errors) [inline compiler/rustc_next_trait_solver/src/solve/eval_ctxt/canonical.rs:415]: This is equivalent, afaict? Why was it using indices previously?

File: compiler/rustc_next_trait_solver/src/solve/eval_ctxt/canonical.rs [hunk 10]
```diff
@@ -425,45 +396,33 @@ where
...
+    {
+        // FIXME: This is so ugly.
```
Comment (compiler-errors) [inline compiler/rustc_next_trait_solver/src/solve/eval_ctxt/canonical.rs:417]: This comment is because this is the only place in the solver that currently uses `Span`, and b/c I had to add a new `fresh_var`-like fn that doesn't have the same shape as the other fns. I could make the wording less useless if desired.

File: compiler/rustc_next_trait_solver/src/solve/eval_ctxt/mod.rs [hunk 11]
```diff
@@ -127,32 +116,69 @@ impl<I: Interner> NestedGoals<I> {
...
-#[extension(pub trait InferCtxtEvalExt<'tcx>)]
-impl<'tcx> InferCtxt<'tcx> {
+pub trait SolverDelegateEvalExt: SolverDelegate {
+    fn evaluate_root_goal(
+        &self,
+        goal: Goal<Self::Interner, <Self::Interner as Interner>::Predicate>,
+        generate_proof_tree: GenerateProofTree,
+    ) -> (Result<(bool, Certainty), NoSolution>, Option<inspect::GoalEvaluation<Self::Interner>>);
+
+    fn evaluate_root_goal_raw(
```
Comment (compiler-errors) [inline compiler/rustc_next_trait_solver/src/solve/eval_ctxt/mod.rs:132]: We have to expose this for use in `analyse.rs` which is still in `rustc_trait_selection`. Why? Because `analyse.rs` uses fulfillment (the `constrain` function uses an `ocx`).

 I could put a FIXME to remove this when fulfillment gets uplifted, though.

File: compiler/rustc_trait_selection/src/solve/infcx.rs [hunk 12]
```diff
@@ -0,0 +1,435 @@
...
+use rustc_middle::ty::{self, Ty, TyCtxt, TypeVisitableExt as _};
+use rustc_span::{ErrorGuaranteed, Span, DUMMY_SP};
+use rustc_type_ir::relate::Relate;
+use rustc_type_ir::solve::{Certainty, NoSolution, SolverMode};
+
+use crate::traits::coherence::trait_ref_is_knowable;
+use crate::traits::specialization_graph;
+
+#[repr(transparent)]
+pub struct SolverDelegate<'tcx>(InferCtxt<'tcx>);
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/solve/infcx.rs:25]: The reason I'm doing this is because I didn't want to something like:

```
struct SolverDelegate<'a, 'tcx>(&'a InferCtxt<'tcx>);
```

because then we'd be going through two layers of indirection.

File: compiler/rustc_type_ir/src/inherent.rs [hunk 13]
```diff
@@ -126,6 +136,49 @@ pub trait Ty<I: Interner<Ty = Self>>:
...
+    fn discriminant_ty(self, interner: I) -> I::Ty;
+
+    fn async_destructor_ty(self, interner: I) -> I::Ty;
+
+    /// Returns `true` when the outermost type cannot be further normalized,
+    /// resolved, or instantiated. This includes all primitive types, but also
+    /// things like ADTs and trait objects, sice even if their arguments or
+    /// nested types may be further simplified, the outermost [`TyKind`] or
+    /// type constructor remains the same.
+    fn is_known_rigid(self) -> bool {
```
Comment (compiler-errors) [inline compiler/rustc_type_ir/src/inherent.rs:149]: Sorry for duplicating this but since it's an exhaustive match, I'm less worried about them getting out of sync.

File: compiler/rustc_next_trait_solver/src/solve/search_graph.rs [hunk 14]
```diff
@@ -465,21 +454,22 @@ enum StepResult<I: Interner> {
...
 
-impl<'tcx> SearchGraph<TyCtxt<'tcx>> {
+impl<I: Interner> SearchGraph<I> {
     /// When we encounter a coinductive cycle, we have to fetch the
     /// result of that cycle while we are still computing it. Because
     /// of this we continuously recompute the cycle until the result
     /// of the previous iteration is equal to the final result, at which
     /// point we are done.
-    fn fixpoint_step_in_task<F>(
+    fn fixpoint_step_in_task<Infcx, F>(
```
Comment (compiler-errors) [inline compiler/rustc_next_trait_solver/src/solve/search_graph.rs:463]: Could make the search graph generic over `Infcx` and use a `PhantomData` just like we do with `ProofTreeBuilder`?

File: compiler/rustc_next_trait_solver/src/solve/search_graph.rs [hunk 15]
```diff
@@ -1,22 +1,19 @@
...
-use super::SolverMode;
-use crate::solve::FIXPOINT_STEP_LIMIT;
+use crate::infcx::SolverDelegate;
+use crate::solve::inspect::{self, ProofTreeBuilder};
+use crate::solve::{
+    CacheData, CanonicalInput, Certainty, QueryResult, SolverMode, FIXPOINT_STEP_LIMIT,
+};
+
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+pub struct SolverLimit(usize);
```
Comment (compiler-errors) [inline compiler/rustc_next_trait_solver/src/solve/search_graph.rs:16]: Ad-hoc reintroduced a `Limit`-like newtype. We should probably consolidate this with the normal `Limit`.

File: compiler/rustc_next_trait_solver/src/solve/normalizes_to/opaque_types.rs [hunk 16]
```diff
@@ -82,10 +86,52 @@ impl<'tcx> EvalCtxt<'_, InferCtxt<'tcx>> {
...
                 self.eq(goal.param_env, expected, actual)?;
                 self.evaluate_added_goals_and_make_canonical_response(Certainty::Yes)
             }
         }
     }
 }
+
+/// Checks whether each generic argument is simply a unique generic placeholder.
+///
+/// This is used in the new solver, which canonicalizes params to placeholders
```
Comment (compiler-errors) [inline compiler/rustc_next_trait_solver/src/solve/normalizes_to/opaque_types.rs:99]: This comment is redundant I guess.

File: compiler/rustc_next_trait_solver/src/solve/normalizes_to/mod.rs [hunk 17]
```diff
@@ -860,93 +865,56 @@ impl<'tcx> assembly::GoalKind<'tcx> for NormalizesTo<'tcx> {
...
-        impl_def_id: DefId,
-    ) -> Result<Option<LeafDef>, NoSolution> {
-        let node_item =
-            specialization_graph::assoc_def(self.interner(), impl_def_id, trait_assoc_def_id)
-                .map_err(|ErrorGuaranteed { .. }| NoSolution)?;
-
-        let eligible = if node_item.is_final() {
-            // Non-specializable items are always projectable.
-            true
+        Ok(if target_container_def_id == impl_trait_ref.def_id {
```
Comment (compiler-errors) [inline compiler/rustc_next_trait_solver/src/solve/normalizes_to/mod.rs:896]: This logic was rewritten to avoid `LeafDef` and `Node`.

File: compiler/rustc_next_trait_solver/src/solve/normalizes_to/mod.rs [hunk 18]
```diff
@@ -137,24 +136,23 @@ impl<'tcx> assembly::GoalKind<'tcx> for NormalizesTo<'tcx> {
...
-        impl_def_id: DefId,
-    ) -> Result<Candidate<TyCtxt<'tcx>>, NoSolution> {
+        ecx: &mut EvalCtxt<'_, Infcx>,
+        goal: Goal<I, NormalizesTo<I>>,
+        impl_def_id: I::DefId,
+    ) -> Result<Candidate<I>, NoSolution> {
         let tcx = ecx.interner();
 
         let goal_trait_ref = goal.predicate.alias.trait_ref(tcx);
-        let impl_trait_header = tcx.impl_trait_header(impl_def_id).unwrap();
```
Comment (compiler-errors) [inline compiler/rustc_next_trait_solver/src/solve/normalizes_to/mod.rs:147]: I split `impl_trait_header` back into `impl_trait_ref` and `impl_polarity` b/c I couldn't be bothered to uplift `ImplHeader`. I can leave a FIXME to add it back if it matters that much.

File: compiler/rustc_middle/src/ty/context.rs [hunk 19]
```diff
@@ -322,6 +371,257 @@ impl<'tcx> Interner for TyCtxt<'tcx> {
             .filter(|assoc_item| matches!(assoc_item.kind, ty::AssocKind::Type))
             .map(|assoc_item| assoc_item.def_id)
     }
+
+    fn args_may_unify_deep(
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/context.rs:375]:
```suggestion
    // FIXME: move `fast_reject` into `rustc_type_ir`.
    fn args_may_unify_deep(
```

File: compiler/rustc_middle/src/ty/generic_args.rs [hunk 20]
```diff
@@ -281,7 +298,8 @@ impl<'tcx> GenericArg<'tcx> {
     pub fn is_non_region_infer(self) -> bool {
         match self.unpack() {
             GenericArgKind::Lifetime(_) => false,
-            GenericArgKind::Type(ty) => ty.is_ty_or_numeric_infer(),
+            // FIXME: This shouldn't return numerical/float.
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/generic_args.rs:301]: it should or it shouldn't? The fixme is either out of date or wrong. I think it's out of date

File: compiler/rustc_next_trait_solver/src/solve/assembly/mod.rs [hunk 21]
```diff
@@ -1,73 +1,70 @@
...
+    BuiltinImplSource, CandidateSource, CanonicalResponse, Certainty, EvalCtxt, Goal, GoalSource,
+    MaybeCause, NoSolution, QueryResult, SolverMode,
+};
+
 /// A candidate is a possible way to prove a goal.
 ///
 /// It consists of both the `source`, which describes how that goal would be proven,
 /// and the `result` when using the given `source`.
-#[derive(Derivative)]
+///
```
Comment (lcnr) [inline compiler/rustc_next_trait_solver/src/solve/assembly/mod.rs:23]:
```suggestion
```

File: compiler/rustc_next_trait_solver/src/solve/eval_ctxt/canonical.rs (hunk 7 above)
Comment (lcnr) [reply compiler/rustc_next_trait_solver/src/solve/eval_ctxt/canonical.rs:100]: please add a FIXME to uplift the leak check instead.

File: compiler/rustc_next_trait_solver/src/solve/eval_ctxt/canonical.rs (hunk 9 above)
Comment (lcnr) [reply compiler/rustc_next_trait_solver/src/solve/eval_ctxt/canonical.rs:415]: nobody knows :shrug:

File: compiler/rustc_next_trait_solver/src/solve/inspect/build.rs [hunk 22]
```diff
@@ -34,9 +24,22 @@ use rustc_type_ir::{self as ty, InferCtxtLike, Interner};
 /// trees. At the end of trait solving `ProofTreeBuilder::finalize`
 /// is called to recursively convert the whole structure to a
 /// finished proof tree.
-pub(in crate::solve) struct ProofTreeBuilder<Infcx, I = <Infcx as InferCtxtLike>::Interner>
+use std::marker::PhantomData;
```
Comment (lcnr) [inline compiler/rustc_next_trait_solver/src/solve/inspect/build.rs:27]: why did you move the imports below the doc comment?

File: compiler/rustc_next_trait_solver/src/solve/normalizes_to/mod.rs [hunk 23]
```diff
@@ -198,21 +194,23 @@ impl<'tcx> assembly::GoalKind<'tcx> for NormalizesTo<'tcx> {
...
                 let error_term = match goal.predicate.alias.kind(tcx) {
                     ty::AliasTermKind::ProjectionTy => Ty::new_error(tcx, guar).into(),
-                    ty::AliasTermKind::ProjectionConst => ty::Const::new_error(tcx, guar).into(),
-                    kind => bug!("expected projection, found {kind:?}"),
+                    ty::AliasTermKind::ProjectionConst => Const::new_error(tcx, guar).into(),
+                    kind => panic!("expected projection, found {kind:?}"),
                 };
                 ecx.instantiate_normalizes_to_term(goal, error_term);
-                ecx.evaluate_added_goals_and_make_canonical_response(Certainty::Yes)
+                return ecx.evaluate_added_goals_and_make_canonical_response(Certainty::Yes);
```
Comment (lcnr) [inline compiler/rustc_next_trait_solver/src/solve/normalizes_to/mod.rs:205]:
```suggestion
                ecx.evaluate_added_goals_and_make_canonical_response(Certainty::Yes)
```
?

File: compiler/rustc_next_trait_solver/src/solve/normalizes_to/opaque_types.rs (hunk 16 above)
Comment (lcnr) [reply compiler/rustc_next_trait_solver/src/solve/normalizes_to/opaque_types.rs:99]: yes

File: compiler/rustc_next_trait_solver/src/solve/normalizes_to/opaque_types.rs [hunk 24]
```diff
@@ -82,10 +86,52 @@ impl<'tcx> EvalCtxt<'_, InferCtxt<'tcx>> {
...
+/// for better caching.
+pub fn uses_unique_placeholders_ignoring_regions<I: Interner>(
+    _interner: I,
```
Comment (lcnr) [inline compiler/rustc_next_trait_solver/src/solve/normalizes_to/opaque_types.rs:102]:
```suggestion
```
or add a FIXME

File: compiler/rustc_next_trait_solver/src/solve/trait_goals.rs [hunk 25]
```diff
@@ -622,22 +613,14 @@ impl<'tcx> assembly::GoalKind<'tcx> for TraitPredicate<'tcx> {
...
         // FIXME: This actually should destructure the `Result` we get from transmutability and
         // register candiates. We probably need to register >1 since we may have an OR of ANDs.
         ecx.probe_builtin_trait_candidate(BuiltinImplSource::Misc).enter(|ecx| {
             let certainty = ecx.is_transmutable(
-                rustc_transmute::Types { dst: args.type_at(0), src: args.type_at(1) },
-                assume,
+                goal.param_env,
+                goal.predicate.trait_ref.args.type_at(0),
+                goal.predicate.trait_ref.args.type_at(1),
+                goal.predicate.trait_ref.args.const_at(2),
```
Comment (lcnr) [inline compiler/rustc_next_trait_solver/src/solve/trait_goals.rs:623]: very ugly, don't really care about that though

File: compiler/rustc_trait_selection/src/solve/fulfill.rs [hunk 26]
```diff
@@ -288,7 +293,10 @@ fn fulfillment_error_for_stalled<'tcx>(
     root_obligation: PredicateObligation<'tcx>,
 ) -> FulfillmentError<'tcx> {
     let (code, refine_obligation) = infcx.probe(|_| {
-        match infcx.evaluate_root_goal(root_obligation.clone().into(), GenerateProofTree::No).0 {
+        match <&SolverDelegate<'tcx>>::from(infcx)
+            .evaluate_root_goal(root_obligation.clone().into(), GenerateProofTree::No)
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/solve/fulfill.rs:297]: This seems very meh, I would prefer to implement `evaluate_root_goal` and `evaluate_root_goal_raw` via an extension trait on `InferCtxt`. Why do we need the `SolverDelegate` super trait on that one?

File: compiler/rustc_type_ir/src/lang_items.rs [hunk 27]
```diff
@@ -1,8 +1,34 @@
 /// Lang items used by the new trait solver. This can be mapped to whatever internal
 /// representation of `LangItem`s used in the underlying compiler implementation.
 pub enum TraitSolverLangItem {
```
Comment (lcnr) [inline compiler/rustc_type_ir/src/lang_items.rs:3]: put this in a tidy alphabetically ordered block or add a FIXME for it

**Review by lcnr — COMMENTED** [review]
nice nice

some nits, then r=me

good work

@bors rollup=never

File: compiler/rustc_next_trait_solver/src/solve/inspect/build.rs (hunk 22 above)
Comment (compiler-errors) [reply compiler/rustc_next_trait_solver/src/solve/inspect/build.rs:27]: Accident, I think. I deleted all of the imports in every file and probably just accidentally added it here.

File: compiler/rustc_trait_selection/src/solve/fulfill.rs (hunk 26 above)
Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/solve/fulfill.rs:297]: because `EvalCtxt::enter_root` requires a `SolverDelegate`
